Add Dockerfile instructions

jashapiro opened the pull request:
### Which issue does this address?

#76

### Briefly describe the scope of the added docs file, including whether you link out to any external docs.

I added some instructions and examples for building Dockerfiles. These include both general instructions and a couple of specific dockerfile examples that people might be able to work from.

This also includes some updates to the Dockerfiles in `hello-r` and `hello-python` that were used for testing.

### Will you need additional visual aids for these docs?

I don't think so

### Any other comments for reviewers?

I am submitting this as a draft, because I wasn't 100% sure about the scope. 
I wrote this with the idea that not all contributors would need to read this, as the mechanics of creating Dockerfiles can be a bit arcane, and is something I expect the Data Lab staff to handle most of the time. Because of that, I didn't get far into syntax or details of Dockerfile creation. However, I wanted to make sure most of the information we would want a contributor or Data Lab staff member would need was present, at least in some working examples.

I also wasn't exactly sure whether we plan to have other docs about Docker (installation & use), but I didn't include much about that here. I thought about adding a section at the end on building and testing, but did not do that quite yet (I am submitting this as a draft for now).

I also have not yet added it to the nav section, for the same reason.

### Author checklist

- [x] The docs page has been added to the `nav` section in `mkdocs.yml`
- [ ] The docs adhere to the style guide [**structure directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)
- [ ] The docs adhere to the style guide [**language directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)
- [ ] The docs adhere to the style guide [**tone directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#tone)
- [ ] The docs follow additional contributing instructions in [the docs contributing file](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/contributing.md)




### Reviewer checklist

Please refer to the [docs style guide](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md) and the [docs contributing guide](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/contributing.md) while reviewing this pull request.

- [ ] Do the docs adhere to the style guide [**structure directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)?
- [ ] Do the docs adhere to the style guide [**language directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#structure)?
- [ ] Do the docs adhere to the style guide [**tone directions**](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/general-style-guide.md#tone)?
- [ ] Do the docs follow additional contributing instructions in [the docs contributing file](https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/docs/contributing.md)?

sjspielman commented on docs/contributing-to-analyses/determining-requirements/determining-software-requirements.md:
````diff
@@ -93,15 +93,16 @@ library(svglite)
 ```
 
-## Managing software dependencies in Python with conda
+## Managing software dependencies in Python and other languages with conda
````
I'm on the fence for "other languages" since it might imply R? I might prefer "other software dependencies" but that's too long I think..

jashapiro replied:
I changed this to just "with conda" and left the better explanation for the text

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+Before making a Docker image, you will likely want to have fully [defined the software environment](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md) that you will be using for your analysis module.
+Most likely, this will take the form of either:
+
+- an [`renv.lock` file](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#determining-and-managing-software-dependencies-in-r) if you are working with R, or
```
"an" 🌶️

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+- an [`renv.lock` file](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#determining-and-managing-software-dependencies-in-r) if you are working with R, or
+- a [conda `environment.yml`](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#module-specific-conda-environments) and/or [`conda-lock.yml` file](../../contributing-to-analyses/determining-requirements/determining-software-requirements.md#freezing-dependencies-with-conda-lock) if you are working with Python or other software available through [Bioconda] ...
+
+The goal of creating a Docker image is to incorporate these software requirements into a container that can be used to run all parts of the analysis module without any installation steps other than installing Docker and pulling the image.
```
I'd move this up to the top of the document, seems like important context for the page overall.

sjspielman replied:
it's kind of long too, so maybe 2 sentence-ify it

jashapiro replied:
I kind of reworked this to have a different emphasis, and moved it.

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+
+The goal of creating a Docker image is to incorporate these software requirements into a container that can be used to run all parts of the analysis module without any installation steps other than installing Docker and pulling the image.
+
+You will also need to have [Docker installed on your computer](https://docs.docker.com/engine/install/) to build and test the images.
```
fyi we have this! https://openscpca.readthedocs.io/en/latest/technical-setup/environment-setup/install-docker/

jashapiro replied:
Changing the link!

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+
+## Docker template files
+
+When you create an analysis module, a template `Dockerfile` and a `.dockerignore` file will be created in the root of the analysis module.
```
I'd link to the analysis modules index page here

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+
+When you create an analysis module, a template `Dockerfile` and a `.dockerignore` file will be created in the root of the analysis module.
+
+The  `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.
```
```suggestion
The `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.
```

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+The  `Dockerfile` includes a `FROM` statement that specifies the base image to use and a couple of `LABEL` statements to populate metadata about the image.
+
+The `.dockerignore` file is used to specify files and directories that should *not* be included in the Docker build environment.
+By default, this will be set to ignore all files with the exception of environment files (e.g., `renv.lock`, `environment.yml`, `conda-lock.yml`).
```
```suggestion
By default, this will be set to ignore all files except environment files (e.g., `renv.lock`, `environment.yml`, `conda-lock.yml`).
```

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+
+## R-based images
+
+If you are working with R, you will likely want to use one of the [`rocker/r-ver` images](https://hub.docker.com/r/rocker/r-ver/tags) or a[`bioconductor/r-ver` image](https://hub.docker.com/r/bioconductor/r-ver/tags) images, which are a set of Docker images that provide complete R environments, tied to a specific version of R and/or Bioconductor.
```
```suggestion
If you are working with R, you will likely want to use one of the [`rocker/r-ver` images](https://hub.docker.com/r/rocker/r-ver/tags) or a [`bioconductor/r-ver` image](https://hub.docker.com/r/bioconductor/r-ver/tags) images, which are a set of Docker images that provide complete R environments, tied to a specific version of R and/or Bioconductor.
```

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+## R-based images
+
+If you are working with R, you will likely want to use one of the [`rocker/r-ver` images](https://hub.docker.com/r/rocker/r-ver/tags) or a[`bioconductor/r-ver` image](https://hub.docker.com/r/bioconductor/r-ver/tags) images, which are a set of Docker images that provide complete R environments, tied to a specific version of R and/or Bioconductor.
+The Bioconductor images are especially handy as they include system software dependencies for Bioconductor packages that may not be installed in the `rocker` images, and support binary installation of Bioconductor packages, which can dramatically speed up build times.
```
```suggestion
The Bioconductor images are especially handy as they include system software dependencies for Bioconductor packages that may not be installed in the `rocker` images.
They also support binary installation of Bioconductor packages, which can dramatically speed up build times.
```

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+- Use `renv::restore()` to reproduce the R environment from the `renv.lock` file.
+
+
+A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might look like this:
```
what about `LABEL`? if it's in the template, it should be here too

jashapiro replied:
I wanted to focus on the mechanics with these examples. The labels are long and distracting.

sjspielman commented on docs/software-platforms/docker/docker-images.md:
````diff
@@ -0,0 +1,114 @@
+
+A simple `Dockerfile` for a conda-based analysis module that uses `conda-lock` for its environment might look like this:
+
+```Dockerfile
````
same `LABEL`  comment

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+- Use `conda-lock install` to create an environment from the `conda-lock.yml` file.
+- Make sure the environment is activated when the container is launched by modifying the `.bashrc` file and setting the `ENTRYPOINT` to `bash -l -c`.
+
+A simple `Dockerfile` for a conda-based analysis module that uses `conda-lock` for its environment might look like this:
```
```suggestion
A simple `Dockerfile` for a conda-based analysis module that uses `conda-lock` for its environment might begin like this:
```

jashapiro replied:
This is the whole thing though? (Aside from labels, as noted)

sjspielman replied:
Ah yeah, I suppose it is. I was thinking about what if more dependencies are needed, but they'd be handled with the conda lines!

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,114 @@
+- Use `renv::restore()` to reproduce the R environment from the `renv.lock` file.
+
+
+A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might look like this:
```
```suggestion
A simple `Dockerfile` for an R-based analysis module that uses `renv` for its environment might begin like this:
```

sjspielman reviewed (commented):
This seems like a good start and scope to me, left some initial feedback!

jashapiro commented on .github/components/dictionary.txt:
```diff
@@ -9,6 +9,8 @@ anonymize
 authenticator
 benchmarking
+Bioconda
+Bioconductor
```
I actually added this prophylactically. Bioconda came up, but not Bioconductor for some reason!

jashapiro commented on analyses/hello-python/Dockerfile:
```diff
@@ -1,7 +1,26 @@
+FROM continuumio/miniconda3:24.1.2-0
 
 # Labels following the Open Containers Initiative (OCI) recommendations
 # For more information, see https://specs.opencontainers.org/image-spec/annotations/?v=v1.0.1
 LABEL org.opencontainers.image.authors="OpenScPCA scpca@ccdatalab.org"
-LABEL org.opencontainers.image.source="https://github.com/AlexsLemonade/OpenScPCA-analysis/tree/main/templates/analysis-module"
+LABEL org.opencontainers.image.source="https://github.com/AlexsLemonade/OpenScPCA-analysis/tree/main/analyses/hello-python"
```
Nope! I didn't mean to delete them there!

jashapiro commented:
@sjspielman As I was adding links to the index, I had the thought that we might want to move the Docker installation instructions to be in the Docker section as well... we probably don't need people to install docker to start, and having all of that together makes some sense to me, but let me know what you think before I start adjusting too many links.

allyhawkins commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,120 @@
+
+This page provides some guidance on creating Docker images for an OpenScPCA analysis module.
+
+Having a Docker image for your analysis module is a key part of ensuring that your analysis is as easily reproducible as possible.
```
```suggestion
Having a Docker image for your analysis module is a key part of ensuring that your analysis is as reproducible as possible.
```

allyhawkins commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,120 @@
+
+## Base images
+
+The base image you choose will depend on the software requirements of your analysis module.
+You should aim to use a publicly available base image that is in active development and well-maintained in order to ensure that your image is secure and up-to-date.
+You should always use a *versioned* base image to ensure that your analysis module is reproducible, i.e., not the one tagged as `latest`.
```
Maybe just add a note here that we provide recommendations on base images to use for R and python below?

allyhawkins commented on docs/technical-setup/environment-setup/install-docker.md:
I am good with moving this to be within the docker folder rather than in the initial environment set up. I think that makes sense.

allyhawkins reviewed (commented):
I just left a few comments, but mostly I agree with you about moving the install instructions, so I would go ahead and do that.

sjspielman commented:
>  I had the thought that we might want to move the Docker installation instructions to be in the Docker section as well... we probably don't need people to install docker to start, and having all of that together makes some sense to me, but let me know what you think before I start adjusting too many links.

Agreed with moving it!

jashapiro commented:
I think this should be ready for another look now. I updated a number of links, including a few that were flagged by some VSCode plugin that I have as having the wrong anchor links. (This is also why I added some named anchors)

I moved the Docker install page to `index.md` because it seemed more general mostly. I did link to the section on installation a couple of times, even though that really just links elsewhere, as I think the other content on the page may be useful to someone following that link.

jashapiro commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,122 @@
+
+Note that creating docker images is not required for contributing to OpenScPCA, but it can be useful for testing and sharing your analysis module with others.
+Each analysis module will ultimately be updated to include a Docker image before producing final results, but this does not need to be done by the module author.
+Data Lab staff will create a Docker image as part of the process of adding the module to the `OpenScPCA-nf` workflow if it has not already been done.
```
We are going to need to add more references to it at some point in the not too distant future. Once we have some downstream analysis that we want available to users, we will want to show people where that comes from.

sjspielman replied:
Ah I see it also in `docs/software-platforms/docker/index.md`, but without much explanation there either.

sjspielman commented on docs/software-platforms/docker/docker-images.md:
````diff
@@ -0,0 +1,122 @@
+```
+
+
+## Conda-based images
````
```suggestion
### Conda-based images
```

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,122 @@
+If your analysis module requires additional software that is not available in the base image, you will need to install it in the Dockerfile.
+The easiest way to do this is likely to be to use the same tool (`renv`, `conda`, etc.) that you used to define the software environment for your analysis module.
+If necessary, you can also include statements to install other software using the system package manager (e.g., `apt-get` for Debian-based images).
+
```
```suggestion

## Base image recommendations

```

sjspielman commented on docs/software-platforms/docker/docker-images.md:
```diff
@@ -0,0 +1,122 @@
+You should aim to use a publicly available base image that is in active development and well-maintained in order to ensure that your image is secure and up-to-date.
+You should always use a *versioned* base image to ensure that your analysis module is reproducible, i.e., not the one tagged as `latest`.
+
+We present some recommended base images for R and conda-based environments below, but you may need to choose a different base image depending on your specific requirements.
```
Something about this organization seemed a smidge off to me, so I suggested some change in headers below, but I'm not entirely sure how I feel about the suggestions. I feel like R/conda-based images should be nested within a base images header, so you could also move those two as `###` sections up here, and then keep `## additional software` as the last header

jashapiro replied:
> Something about this organization seemed a smidge off to me, so I suggested some change in headers below, but I'm not entirely sure how I feel about the suggestions. I feel like R/conda-based images should be nested within a base images header, so you could also move those two as `###` sections up here, and then keep `## additional software` as the last header

I updated the organization a little bit differently, making two main headers and putting these under an "Examples" header. Let me know what you think

sjspielman replied:
I like this a lot!

sjspielman commented on docs/software-platforms/docker/index.md:
```diff
@@ -1,30 +1,30 @@
+
+We will also use Docker images to run the analysis modules in the `OpenScPCA-nf` workflow when generating final analysis results.
 
 ## How to install Docker
```
here - https://github.com/AlexsLemonade/OpenScPCA-analysis/issues/398

jashapiro commented:
> No major comments except for figuring out the best way to introduce `OpenScPCA-nf`.

I think we should leave that for a different PR?

sjspielman commented:
> > No major comments except for figuring out the best way to introduce `OpenScPCA-nf`.
> 
> I think we should leave that for a different PR?

Yeah, it'll be too much for this one. Let's make sure we have an issue though!

sjspielman reviewed (approved):
I like the re-organization! Everything else looks good at this point, let's just open an openscpca-nf docs issue.
